hw/arm/guest-socket.c

```diff
@@ -40,6 +40,74 @@ static int32_t find_free_socket(void) {
     return -1;
 }
 
+
+/* NETWORK_HACK:
+    See `guest-services/socket.h`:
+    Darwin still uses `sin_len` in sockaddr,
+     causing errors when copying memory from iOS.
+*/
+#ifdef __APPLE__
+# define ADDR addr
+# undef NETWORK_HACK
+#else /* Linux, etc. */
+# define ADDR darwin_addr
+# define NETWORK_HACK 1
+#endif
+
+static int convert_error(int err) {
+#ifndef NETWORK_HACK
+     return err;
+#else /* linux */
+    int result;
+
+    switch(err) {
+        case EAGAIN:
+/*      case EWOULDBLOCK: */
+            result = 35;
+            break;
+        case EDEADLK:
+/*      case EDEADLOCK: */
+            result = 11;
+            break;
+        /* case {...}: */
+        default:
+            result = err;
+            break;
+   }
+   return result;
+#endif
+}
+
+static struct sockaddr_in convert_sockaddr_to(S_SOCKADDR_IN from) {
+#ifndef NETWORK_HACK
+    return from;
+#else
+    struct sockaddr_in to;
+    to.sin_family   = from.sin_family;
+    to.sin_port     = from.sin_port;
+    to.sin_addr     = from.sin_addr;
+    memset(&to.sin_zero, 0, sizeof(to.sin_zero));
+
+    return to;
+#endif
+}
+
+static S_SOCKADDR_IN convert_sockaddr_from(struct sockaddr_in from) {
+#ifndef NETWORK_HACK
+    return from;
+#else
+    S_SOCKADDR_IN to;
+    to.sin_family   = from.sin_family;
+    to.sin_port     = from.sin_port;
+    to.sin_addr     = from.sin_addr;
+    to.sin_len      = sizeof(to);
+    memset(&to.sin_zero, 0, sizeof(to.sin_zero));
+
+    return to;
```

Comment:
Same as above...

```suggestion
    return (darwin_sockaddr_in) {
        sizeof(struct darwin_sockaddr_in),
        from.sin_family,
        from.sin_port,
        from.sin_addr,
        { [0 ... sizeof(from.sin_zero) - 1] = 0 } };
```

I know I'm using `sizeof(from.sin_zero)` instead of `sizeof(to.sin_zero)`, but luckily, those sizes are the same.